semantic-analysis/src/Analysis/Syntax/Python.hs

```diff
@@ -1,82 +1,122 @@
-{-# LANGUAGE DeriveTraversable #-}
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE LambdaCase #-}
 -- | This belongs in @semantic-python@ instead of @semantic-analysis@, but for the sake of expedience…
 module Analysis.Syntax.Python
 ( -- * Syntax
-  Term
-, Python(..)
+  Term(..)
+, subterms
   -- * Abstract interpretation
 , eval0
 , eval
+  -- * Parsing
+, parse
 ) where
 
 import           Analysis.Effect.Domain hiding ((:>>>))
 import qualified Analysis.Effect.Statement as S
 import           Analysis.Name
 import           Analysis.Reference
-import qualified Analysis.Syntax as T
 import           Analysis.VM
 import           Control.Effect.Labelled
 import           Control.Effect.Reader
+import           Control.Monad (foldM)
+import           Data.Foldable (for_)
 import           Data.Function (fix)
-import           Data.List.NonEmpty (NonEmpty)
-import           Data.Text (Text)
-import           Source.Span (Span)
+import           Data.List.NonEmpty (nonEmpty)
+import           Data.Maybe (mapMaybe)
+import qualified Data.Set as Set
+import           Data.Text (pack)
+import qualified Language.Python.Common as Py
+import           Language.Python.Version3.Parser
+import           Source.Span (Pos (..), Span (..), point)
+import           System.FilePath (takeBaseName)
 
 -- Syntax
 
-type Term = T.Term Python Name
+data Term
+  = Module (Py.Module Py.SrcSpan)
+  | Statement (Py.Statement Py.SrcSpan)
+  | Expr (Py.Expr Py.SrcSpan)
+  deriving (Eq, Ord, Show)
```

Comment:
We wrap `language-python` AST nodes up in constructors of a single datatype so as to avoid having to define our _own_ AST for Python and a complicated and error-prone function for copying things from the `language-python` AST in. This type isn't recursive at all; we just wrap the current layer, and then wrap the next layer on the fly in recursive functions like `subterms` and `eval`.

Somewhat to my surprise, the `language-python` AST (and, I suppose, Python's own syntax) is regular enough to avoid making this very much more complex. There are lots of other types representing identifiers, arguments, parameters, and all the other, weirder bits of syntax, but most of them are auxiliary—not things we would evaluate.